crates/red_knot_python_semantic/src/types/mro.rs

```diff
@@ -0,0 +1,641 @@
+    /// runtime (and/or that the class creation could possibly fail
+    /// altogether).
+    fn possibly_many(
+        mut possibilities: FxHashSet<Mro<'db>>,
+        mut errors: FxHashSet<BasesList<'db>>,
+    ) -> Self {
+        debug_assert_ne!(
+            possibilities.len().saturating_add(errors.len()),
+            0,
+            "There should always be at least one possible mro outcome"
+        );
+        match (possibilities.len(), errors.len()) {
+            (1, 0) => Self::SingleSuccess(possibilities.into_iter().next().unwrap()),
+            (_, 0) => {
+                possibilities.shrink_to_fit();
+                Self::MultipleSuccesses(possibilities)
+            }
+            (0, _) => {
+                errors.shrink_to_fit();
+                Self::CertainFailure(errors)
+            }
+            _ => {
+                possibilities.shrink_to_fit();
+                errors.shrink_to_fit();
+                Self::PossibleSuccess {
+                    possible_mros: possibilities,
+                    failure_cases: errors,
+                }
+            }
+        }
+    }
+}
+
+/// Fast path for calculating a class's MRO.
+///
+/// This fast path is only valid if we know that none of the bases is a
+/// [`Type::Union`]
+fn mro_of_class_fast_path<'db>(
+    db: &'db dyn Db,
+    class: ClassType<'db>,
+    bases: &[Type<'db>],
+) -> Option<MroPossibilities<'db>> {
+    match bases {
+        // 0 bases means that it must be `object` itself.
+        //
+        // The case for `object` itself isn't really that common,
+        // but we may as well handle it here, since it's known and easy:
+        [] => {
+            debug_assert_eq!(
+                Type::Class(class),
+                KnownClass::Object.to_class(db),
+                "Only `object` should have 0 bases in Python"
+            );
+            Some(MroPossibilities::single([class]))
+        }
+
+        // The class has a single base.
+        //
+        // That could be an explicit base (`class A(B): pass`),
+        // or an implicit base (which is always `object`: `class A: pass`).
+        [single_base] => {
+            let object = KnownClass::Object.to_class(db);
+            let mro = if single_base == &object {
+                MroPossibilities::single([class, object.expect_class()])
+            } else {
+                let mut possibilities = FxHashSet::default();
+                let base = ClassBase::from(single_base);
+                for possibility in base.mro_possibilities(db).iter(db, base) {
+                    possibilities.insert(
+                        std::iter::once(ClassBase::Class(class))
+                            .chain(possibility.iter().copied())
+                            .collect(),
+                    );
+                }
+                MroPossibilities::possibly_many(possibilities, FxHashSet::default())
+            };
```

Comment:
You can avoid the `expect` call by using
```suggestion
            let mro = match single_base.into_class_type() {
                Some(class) if class.is_known(db, KnownClass::Object) => {
                    MroPossibilities::single([class])
                }
                _ => {
                    let mut possibilities = FxHashSet::default();
                    let base = ClassBase::from(single_base);
                    for possibility in base.mro_possibilities(db).iter(db, base) {
                        possibilities.insert(
                            std::iter::once(ClassBase::Class(class))
                                .chain(possibility.iter().copied())
                                .collect(),
                        );
                    }
                    MroPossibilities::possibly_many(possibilities, FxHashSet::default())
                }
            };
```

Reply:
Very nice!